lib/BridgeInterface.php

```diff
@@ -143,4 +143,8 @@ public function getCacheTimeout();
      * @return array|null List of bridge parameters or null if detection failed.
      */
     public function detectParameters($url);
+
+    public function isDeprecated(): bool;
```

Comment:
As before, I would still argue for separate `DeprecatedBridge` interface. Then `$bridge->isDeprecated()` could be replaced by `$bridge instanceof DeprecatedBridge`, adding `isDeprecated()` method would be replaced by `implements DeprecatedBridge` and it would be impossible to forget to add the deprecation message. Also it would not add more clutter to the `AbstractBridge`.

Reply:
Also there would be no need for any exceptions and the deprecation would be detectable without instantiating the bridge.

Reply:
I think your solution is slightly cleaner. I would still put the `isDeprecated()` method inside `AbstractBridge` for encapsulation. The IDE will not find the `getDeprecationMessage()` method here since its typehint is `BridgeInterface`.

```php
if ($bridge instanceof DeprecatedBridge) {
  print $bridge->getDeprecationMessage(); // IDE says unknown method but we know it exists
}
```

How would you solve this?

Reply:
Is that just not being able to suggest it, or will it show annoying error? Either way that sounds like a bug with the IDE, if it cannot detect the method within a conditional, so I would report it to the IDE developers and ignore it.